File: src/discord/embed.ts
import type { ChangeType } from "../config/schema";
import type { DetectedChange } from "../monitor/detector";

/**
 * @description タイトル/ゲーム変更イベントの種別一覧
 */
const CHANGE_EVENT_TYPES: ReadonlySet<ChangeType> = new Set([
  "titleChange",
  "gameChange",
  "titleAndGameChange",
]);

/**
 * @description Discord Embed構造
 * @property title - Embedのタイトル
 * @property description - Embedの説明 @optional
 * @property url - クリック時に開くURL @optional
 * @property color - 左側の縦線の色(10進数)
 * @property thumbnail - サムネイル画像 @optional
 * @property image - メイン画像 @optional
 * @property fields - フィールドの配列 @optional
 * @property timestamp - タイムスタンプ(ISO 8601) @optional
 * @property footer - フッター情報 @optional
 * @property author - 作成者情報 @optional
 */
export interface DiscordEmbed {
  title: string;
  description?: string;
  url?: string;
  color: number;
  thumbnail?: { url: string };
  image?: { url: string };
  fields?: { name: string; value: string; inline?: boolean }[];
  timestamp?: string;
  footer?: { text: string };
  author?: { name: string; icon_url?: string };
}

/**
 * @description 変更種別ごとのEmbed色(10進数)
 */
const COLORS: Record<ChangeType, number> = {
  online: 0x9146ff,
  offline: 0x808080,
  titleChange: 0x00ff00,
  gameChange: 0xff9900,
  titleAndGameChange: 0x00ccff,
};

/**
 * @description 変更種別ごとのEmbedタイトル
 */
const TITLES: Record<ChangeType, string> = {
  online: "配信開始",
  offline: "配信終了",
  titleChange: "タイトル変更",
  gameChange: "ゲーム変更",
  titleAndGameChange: "タイトル・ゲーム変更",
};

/**
 * @description 配信開始からの経過時間を日本語でフォーマット
 * @param startedAt - 配信開始日時(ISO 8601)
 * @returns フォーマット済み文字列(無効な場合null)
 */
function formatElapsedTime(startedAt: string | null): string | null {
  if (!startedAt) return null;

  const start = new Date(startedAt).getTime();
  const now = Date.now();
  const diffMs = now - start;

  if (diffMs < 0) return null;

  const minutes = Math.floor(diffMs / 60000);
  if (minutes < 1) return "たった今";

  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;

  if (hours === 0) return `${mins}分前から配信中`;
  return `${hours}時間${mins}分前から配信中`;
}

/**
 * @description 配信時間を日本語でフォーマット
 * @param startedAt - 配信開始日時(ISO 8601)
 * @returns フォーマット済み文字列
 */
function formatDuration(startedAt: string): string {
  const start = new Date(startedAt).getTime();
  const now = Date.now();
  const diffMs = now - start;

  const totalMinutes = Math.floor(diffMs / 60000);
  const hours = Math.floor(totalMinutes / 60);
  const mins = totalMinutes % 60;

  if (hours === 0) return `${mins}分`;
  return `${hours}時間${mins}分`;
}

/**
 * @description 変更情報からDiscord Embedを構築
 * @param change - 検出された変更
 * @returns Discord Embed
 */
export function buildEmbed(change: DetectedChange): DiscordEmbed {
  const { type, currentState } = change;
  const channelUrl = `https://twitch.tv/${currentState.username}`;

  const embed: DiscordEmbed = {
    title: TITLES[type],
    url: channelUrl,
    color: COLORS[type],
    author: {
      name: currentState.displayName,
      icon_url: currentState.profileImageUrl,
    },
    timestamp: new Date().toISOString(),
  };

  switch (type) {
    case "online": {
      embed.description = currentState.title || "(タイトルなし)";
      const fields: { name: string; value: string; inline?: boolean }[] = [
        {
          name: "ゲーム",
          value: currentState.gameName || "(未設定)",
          inline: true,
        },
      ];

      if (currentState.startedAt) {
        const startTime = new Date(currentState.startedAt);
        fields.push({
          name: "開始時刻",
          value: startTime.toLocaleTimeString("ja-JP", { hour: "2-digit", minute: "2-digit" }),
          inline: true,
        });

        const elapsed = formatElapsedTime(currentState.startedAt);
        if (elapsed && !elapsed.includes("たった今")) {
          embed.footer = { text: elapsed };
        }
      }

      embed.fields = fields;
      if (currentState.thumbnailUrl) {
        const thumbnailUrl = currentState.thumbnailUrl
          .replace("{width}", "440")
          .replace("{height}", "248");
        embed.image = { url: thumbnailUrl };
      }
      break;
    }

    case "offline": {
      const endTime = new Date().toLocaleTimeString("ja-JP", {
        hour: "2-digit",
        minute: "2-digit",
      });
      const fields: { name: string; value: string; inline?: boolean }[] = [];

      if (change.streamStartedAt) {
        const startTime = new Date(change.streamStartedAt).toLocaleTimeString("ja-JP", {
          hour: "2-digit",
          minute: "2-digit",
        });
        const duration = formatDuration(change.streamStartedAt);
        fields.push({
          name: "配信時間",
          value: `${startTime} → ${endTime} (${duration})`,
          inline: false,
        });
      } else {
        fields.push({ name: "終了時刻", value: endTime, inline: true });
      }

      if (change.vodUrl) {
        fields.push({
          name: "VOD",
          value: `[この配信を見る](${change.vodUrl})`,
          inline: false,
        });
      }

      embed.description = "配信が終了しました";
      embed.fields = fields;

      if (change.vodThumbnailUrl) {
        embed.image = { url: change.vodThumbnailUrl };
      }
      break;
    }

    case "titleChange":
      embed.fields = [
        { name: "変更前", value: change.oldValue || "(なし)", inline: false },
        { name: "変更後", value: change.newValue || "(なし)", inline: false },
      ];
      break;

    case "gameChange":
      embed.fields = [
        { name: "変更前", value: change.oldValue || "(未設定)", inline: true },
        { name: "変更後", value: change.newValue || "(未設定)", inline: true },
      ];
      break;

    case "titleAndGameChange":
      embed.fields = [
        {
          name: "タイトル",
          value: `${change.oldTitle || "(なし)"}\n→ ${change.newTitle || "(なし)"}`,
          inline: false,
        },
        {
          name: "ゲーム",
          value: `${change.oldGame || "(未設定)"}\n→ ${change.newGame || "(未設定)"}`,
          inline: false,
        },
      ];
      break;
  }

  // タイトル/ゲーム変更時は配信中であればfooterを設定
  if (CHANGE_EVENT_TYPES.has(type) && currentState.isLive && !embed.footer) {
    embed.footer = { text: "配信中" };
  }

  return embed;
}

File: internal/discord/embed.go
// Package discord はDiscord Webhook連携を提供する。
package discord

import (
	"fmt"
	"strings"
	"time"

	"github.com/yuu1111/StreamNotifier/internal/config"
	"github.com/yuu1111/StreamNotifier/internal/monitor"
)

// EmbedField はDiscord Embedのフィールド。
type EmbedField struct {
	Name   string `json:"name"`
	Value  string `json:"value"`
	Inline bool   `json:"inline,omitempty"`
}

// EmbedFooter はDiscord Embedのフッター。
type EmbedFooter struct {
	Text string `json:"text"`
}

// EmbedAuthor はDiscord Embedの作成者情報。
type EmbedAuthor struct {
	Name    string `json:"name"`
	IconURL string `json:"icon_url,omitempty"`
}

// EmbedImage はDiscord Embedの画像。
type EmbedImage struct {
	URL string `json:"url"`
}

// Embed はDiscord Embed構造。
type Embed struct {
	Title       string       `json:"title"`
	Description string       `json:"description,omitempty"`
	URL         string       `json:"url,omitempty"`
	Color       int          `json:"color"`
	Thumbnail   *EmbedImage  `json:"thumbnail,omitempty"`
	Image       *EmbedImage  `json:"image,omitempty"`
	Fields      []EmbedField `json:"fields,omitempty"`
	Timestamp   string       `json:"timestamp,omitempty"`
	Footer      *EmbedFooter `json:"footer,omitempty"`
	Author      *EmbedAuthor `json:"author,omitempty"`
}

var colorMap = map[string]int{
	config.ChangeOnline:       0x9146ff,
	config.ChangeOffline:      0x808080,
	config.ChangeTitleChange:  0x00ff00,
	config.ChangeGameChange:   0xff9900,
	config.ChangeTitleAndGame: 0x00ccff,
}

var titleMap = map[string]string{
	config.ChangeOnline:       "配信開始",
	config.ChangeOffline:      "配信終了",
	config.ChangeTitleChange:  "タイトル変更",
	config.ChangeGameChange:   "ゲーム変更",
	config.ChangeTitleAndGame: "タイトル・ゲーム変更",
}

// changeEventTypes はタイトル/ゲーム変更系のイベント種別。
var changeEventTypes = map[string]bool{
	config.ChangeTitleChange:  true,
	config.ChangeGameChange:   true,
	config.ChangeTitleAndGame: true,
}

// formatElapsedTime は配信開始からの経過時間を日本語でフォーマットする。
func formatElapsedTime(startedAt string) string {
	start, err := time.Parse(time.RFC3339, startedAt)
	if err != nil {
		return ""
	}

	diff := time.Since(start)
	if diff < 0 {
		return ""
	}

	totalMinutes := int(diff.Minutes())
	if totalMinutes < 1 {
		return "たった今"
	}

	hours := totalMinutes / 60
	mins := totalMinutes % 60

	if hours == 0 {
		return fmt.Sprintf("%d分前から配信中", mins)
	}
	return fmt.Sprintf("%d時間%d分前から配信中", hours, mins)
}

// formatDuration は配信時間をフォーマットする。
func formatDuration(startedAt string) string {
	start, err := time.Parse(time.RFC3339, startedAt)
	if err != nil {
		return "不明"
	}

	diff := time.Since(start)
	totalMinutes := int(diff.Minutes())
	hours := totalMinutes / 60
	mins := totalMinutes % 60

	if hours == 0 {
		return fmt.Sprintf("%d分", mins)
	}
	return fmt.Sprintf("%d時間%d分", hours, mins)
}

// formatTimeJST は時刻をJST HH:MM形式にフォーマットする。
func formatTimeJST(t time.Time) string {
	jst := time.FixedZone("JST", 9*60*60)
	return t.In(jst).Format("15:04")
}

// orDefault は空文字列の場合にデフォルト値を返す。
func orDefault(s, defaultVal string) string {
	if s == "" {
		return defaultVal
	}
	return s
}

// BuildEmbed は変更情報からDiscord Embedを構築する。
func BuildEmbed(change monitor.DetectedChange) Embed {
	state := change.CurrentState
	channelURL := "https://twitch.tv/" + state.Username

	embed := Embed{
		Title:     titleMap[change.Type],
		URL:       channelURL,
		Color:     colorMap[change.Type],
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Author: &EmbedAuthor{
			Name:    state.DisplayName,
			IconURL: state.ProfileImageURL,
		},
	}

	switch change.Type {
	case config.ChangeOnline:
		embed.Description = orDefault(state.Title, "(タイトルなし)")

		fields := []EmbedField{
			{Name: "ゲーム", Value: orDefault(state.GameName, "(未設定)"), Inline: true},
		}

		if state.StartedAt != "" {
			startTime, err := time.Parse(time.RFC3339, state.StartedAt)
			if err == nil {
				fields = append(fields, EmbedField{
					Name:   "開始時刻",
					Value:  formatTimeJST(startTime),
					Inline: true,
				})

				elapsed := formatElapsedTime(state.StartedAt)
				if elapsed != "" && !strings.Contains(elapsed, "たった今") {
					embed.Footer = &EmbedFooter{Text: elapsed}
				}
			}
		}

		embed.Fields = fields

		if state.ThumbnailURL != "" {
			thumbnailURL := strings.ReplaceAll(state.ThumbnailURL, "{width}", config.ThumbnailWidth)
			thumbnailURL = strings.ReplaceAll(thumbnailURL, "{height}", config.ThumbnailHeight)
			embed.Image = &EmbedImage{URL: thumbnailURL}
		}

	case config.ChangeOffline:
		embed.Description = "配信が終了しました"

		var fields []EmbedField
		now := time.Now()

		if change.StreamStartedAt != "" {
			startTime, err := time.Parse(time.RFC3339, change.StreamStartedAt)
			if err == nil {
				duration := formatDuration(change.StreamStartedAt)
				fields = append(fields, EmbedField{
					Name:  "配信時間",
					Value: fmt.Sprintf("%s → %s (%s)", formatTimeJST(startTime), formatTimeJST(now), duration),
				})
			} else {
				fields = append(fields, EmbedField{
					Name:   "終了時刻",
					Value:  formatTimeJST(now),
					Inline: true,
				})
			}
		} else {
			fields = append(fields, EmbedField{
				Name:   "終了時刻",
				Value:  formatTimeJST(now),
				Inline: true,
			})
		}

		if change.VodURL != "" {
			fields = append(fields, EmbedField{
				Name:  "VOD",
				Value: fmt.Sprintf("[この配信を見る](%s)", change.VodURL),
			})
		}

		embed.Fields = fields

		if change.VodThumbnailURL != "" {
			embed.Image = &EmbedImage{URL: change.VodThumbnailURL}
		}

	case config.ChangeTitleChange:
		embed.Fields = []EmbedField{
			{Name: "変更前", Value: orDefault(change.OldValue, "(なし)")},
			{Name: "変更後", Value: orDefault(change.NewValue, "(なし)")},
		}

	case config.ChangeGameChange:
		embed.Fields = []EmbedField{
			{Name: "変更前", Value: orDefault(change.OldValue, "(未設定)"), Inline: true},
			{Name: "変更後", Value: orDefault(change.NewValue, "(未設定)"), Inline: true},
		}

	case config.ChangeTitleAndGame:
		embed.Fields = []EmbedField{
			{
				Name:  "タイトル",
				Value: fmt.Sprintf("%s\n→ %s", orDefault(change.OldTitle, "(なし)"), orDefault(change.NewTitle, "(なし)")),
			},
			{
				Name:  "ゲーム",
				Value: fmt.Sprintf("%s\n→ %s", orDefault(change.OldGame, "(未設定)"), orDefault(change.NewGame, "(未設定)")),
			},
		}
	}

	// タイトル/ゲーム変更時は配信中であればfooterを設定
	if changeEventTypes[change.Type] && state.IsLive && embed.Footer == nil {
		embed.Footer = &EmbedFooter{Text: "配信中"}
	}

	return embed
}
